File: leetcode/medium/evaluate-reverse-polish-notation.cpp
// https://leetcode.com/problems/evaluate-reverse-polish-notation/
//
// Complexity: runtime O(n), space O(n).

class Solution {
public:
    int evalRPN(const vector<string>& tokens) const
    {
        stack<int> s;

        for (const auto& t : tokens) {
            if (int val; istringstream(t) >> val) {
                s.push(val); // valid integer
            } else {
                const int b = s.top();
                s.pop();

                const int a = s.top();
                s.pop();

                if (t == "+")
                    s.push(a + b);
                else if (t == "-")
                    s.push(a - b);
                else if (t == "*")
                    s.push(a * b);
                else if (t == "/")
                    s.push(a / b);
            }
        }

        return !empty(s) ? s.top() : 0;
    }
};

File: leetcode/medium/evaluate-reverse-polish-notation.rs
// https://leetcode.com/problems/evaluate-reverse-polish-notation
//
// Complexity: runtime O(n), space O(n).

impl Solution {
    pub fn eval_rpn(tokens: Vec<String>) -> i32 {
        let mut s = vec![];
        for t in &tokens {
            if let Ok(x) = t.parse() {
                s.push(x);
            } else {
                let (y, x) = (s.pop().unwrap(), s.pop().unwrap());
                s.push(match t.as_str() {
                    "+" => x + y,
                    "-" => x - y,
                    "*" => x * y,
                    "/" => x / y,
                    _ => unreachable!(),
                });
            }
        }
        s.last().copied().unwrap_or(0)
    }
}
